### PR TITLE
Make local_driver kill() robust wrt timing

### LINKED ISSUE #7704: ProcessLookupError when terminating a local ensemble
This line:
https://github.com/equinor/ert/blob/8548bcd685e05bed0b5b6464fca5595c48ce6bea/src/ert/scheduler/local_driver.py#L97

can trigger an error if a job has been initiated by `local_driver.py` but the subprocess has not yet started. Observed while bug-bashing:

```
Killing realization 71 failed with error 
Exception in scheduler task job-59_task: 
Traceback: Traceback (most recent call last):
  File "/private/havb/projects/ert/src/ert/scheduler/local_driver.py", line 72, in _run
    returncode = await self._wait(proc)
  File "/private/havb/projects/ert/src/ert/scheduler/local_driver.py", line 91, in _wait
    return await proc.wait()
  File "/opt/rh/rh-python38/root/usr/lib64/python3.8/asyncio/subprocess.py", line 135, in wait
    return await self._transport._wait()
  File "/opt/rh/rh-python38/root/usr/lib64/python3.8/asyncio/base_subprocess.py", line 235, in _wait
    return await waiter
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/private/havb/projects/ert/src/ert/scheduler/job.py", line 131, in run
    await self._submit_and_run_once(sem)
  File "/private/havb/projects/ert/src/ert/scheduler/job.py", line 119, in _submit_and_run_once
    await self.driver.kill(self.iens)
  File "/private/havb/projects/ert/src/ert/scheduler/local_driver.py", line 47, in kill
    raise err
  File "/private/havb/projects/ert/src/ert/scheduler/local_driver.py", line 40, in kill
    await self._tasks[iens]
  File "/opt/rh/rh-python38/root/usr/lib64/python3.8/asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File "/private/havb/projects/ert/src/_ert/async_utils.py", line 53, in _done_callback
    raise exc
  File "/private/havb/projects/ert/src/ert/scheduler/local_driver.py", line 76, in _run
    returncode = await self._kill(proc)
  File "/private/havb/projects/ert/src/ert/scheduler/local_driver.py", line 97, in _kill
    proc.terminate()
  File "/opt/rh/rh-python38/root/usr/lib64/python3.8/asyncio/subprocess.py", line 141, in terminate
    self._transport.terminate()
  File "/opt/rh/rh-python38/root/usr/lib64/python3.8/asyncio/base_subprocess.py", line 149, in terminate
    self._check_proc()
  File "/opt/rh/rh-python38/root/usr/lib64/python3.8/asyncio/base_subprocess.py", line 142, in _check_proc
    raise ProcessLookupError()
ProcessLookupError
```

### PR DESCRIPTION
**Pull request by berland**
If the kill() instruction to the local driver happened before the subprocess had actually started, the driver would raise an exception.

This commit makes sure that a FinishedEvent is always sent no matter when the kill() instruction is sent.

kill() can happen before the _run() function is started, it can happen when run() is started and the try block is entered, but _init() is not started, and it can happen after the process has been started.

**Issue**
Resolves #7704 

**Approach**
_Short description of the approach_

(Screenshot of new behavior in GUI if applicable)

- [ ] PR title captures the intent of the changes, and is fitting for release notes.
- [ ] Added appropriate release note label
- [ ] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [ ] Make sure tests pass locally (after every commit!)

## When applicable
- [ ] **When there are user facing changes**: Updated documentation
- [ ] **New behavior or changes to existing untested code**: Ensured that unit tests are added (See [Ground Rules](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md#ground-rules)).
- [ ] **Large PR**: Prepare changes in small commits for more convenient review
- [ ] **Bug fix**: Add regression test for the bug
- [ ] **Bug fix**: Create Backport PR to latest release

### CODE REVIEW

File: src/ert/scheduler/local_driver.py [hunk 1]
```diff
@@ -39,6 +41,12 @@ async def kill(self, iens: int) -> None:
             with contextlib.suppress(asyncio.CancelledError):
                 await self._tasks[iens]
             del self._tasks[iens]
+            if iens not in self._sent_finished_events:
```
[the remark this answers is not in the document]
Comment (xjules) [reply src/ert/scheduler/local_driver.py:44]: The function name is a suggestion :)

File: src/ert/scheduler/local_driver.py [hunk 2]
```diff
@@ -87,17 +99,18 @@ async def _init(iens: int, executable: str, /, *args: str) -> Process:
...
 
     @staticmethod
     async def _kill(proc: Process) -> int:
-        """This method exists to allow for mocking it in tests"""
         try:
             proc.terminate()
             await asyncio.wait_for(proc.wait(), _TERMINATE_TIMEOUT)
         except asyncio.TimeoutError:
             proc.kill()
+        except ProcessLookupError:
```
Comment (xjules) [inline src/ert/scheduler/local_driver.py:111]: nice!

File: src/ert/scheduler/local_driver.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (berland) [reply src/ert/scheduler/local_driver.py]: Refactored.

File: src/ert/scheduler/local_driver.py [hunk 3]
```diff
@@ -64,17 +68,23 @@ async def _run(self, iens: int, executable: str, /, *args: str) -> None:
...
         try:
             returncode = await self._wait(proc)
             logger.debug(f"Realization {iens} finished with {returncode=}")
-            await self.event_queue.put(FinishedEvent(iens=iens, returncode=returncode))
+            await self._dispatch_finished_event(iens, returncode)
         except asyncio.CancelledError:
             returncode = await self._kill(proc)
+            await self._dispatch_finished_event(iens, returncode)
+
+    async def _dispatch_finished_event(self, iens: int, returncode: int):
```
Comment (xjules) [inline src/ert/scheduler/local_driver.py:83]: missing `-> None:`

File: src/ert/scheduler/local_driver.py [hunk 4]
```diff
@@ -64,17 +68,23 @@ async def _run(self, iens: int, executable: str, /, *args: str) -> None:
...
-            await self.event_queue.put(FinishedEvent(iens=iens, returncode=127))
+            await self._dispatch_finished_event(iens, 127)
             return
 
         await self.event_queue.put(StartedEvent(iens=iens))
         try:
             returncode = await self._wait(proc)
             logger.debug(f"Realization {iens} finished with {returncode=}")
-            await self.event_queue.put(FinishedEvent(iens=iens, returncode=returncode))
+            await self._dispatch_finished_event(iens, returncode)
```
Comment (xjules) [inline src/ert/scheduler/local_driver.py:78]: Now I wonder:
```python
try:
    returncode = await self._wait(proc)
except asyncio.CancelledError:
     returncode = await self._kill(proc)
finally:
   await self._dispatch_finished_event(iens, returncode)
```

**Review by xjules — APPROVED** [review]
Nice @berland ! :rocket:

File: src/ert/scheduler/local_driver.py [hunk 5]
```diff
@@ -64,17 +68,25 @@ async def _run(self, iens: int, executable: str, /, *args: str) -> None:
...
+
+        returncode = 0
         try:
             returncode = await self._wait(proc)
             logger.debug(f"Realization {iens} finished with {returncode=}")
-            await self.event_queue.put(FinishedEvent(iens=iens, returncode=returncode))
         except asyncio.CancelledError:
             returncode = await self._kill(proc)
+        finally:
+            await self._dispatch_finished_event(iens, returncode)
```
Comment (xjules) [inline src/ert/scheduler/local_driver.py:83]: Maybe there should be another ?
```python
except Exception:
    returncode = signal.SIGTERM + SIGNAL_OFFSET
```

**Comment by berland** [comment]
Test errors for max_submit was due to `submit()` now having to do cleanup in `_sent_finished_events`.

File: src/ert/scheduler/local_driver.py (hunk 5 above)
Comment (berland) [reply src/ert/scheduler/local_driver.py:86]: If anything else than CancelledError occurs here, we should crash (for the moment at least)

File: src/ert/scheduler/local_driver.py [hunk 6]
```diff
@@ -31,6 +34,8 @@ async def submit(
         runpath: Optional[Path] = None,
     ) -> None:
         self._tasks[iens] = asyncio.create_task(self._run(iens, executable, *args))
+        with suppress(KeyError):
```
Comment (xjules) [inline src/ert/scheduler/local_driver.py:37]: Nice!

**Pull request merged**